remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java

```diff
@@ -236,15 +292,68 @@ private NodeRef changeCategoryName(final NodeRef categoryNodeRef, final String n
     }
 
     /**
-     * Verify if fixed category name is not empty.
+     * Validate if fixed category name is not empty.
      *
      * @param fixedCategoryModel Fixed category model.
      */
-    private void verifyCategoryFields(final Category fixedCategoryModel)
+    private void validateCategoryFields(final Category fixedCategoryModel)
     {
         if (StringUtils.isEmpty(fixedCategoryModel.getName()))
         {
             throw new InvalidArgumentException(NOT_NULL_OR_EMPTY);
         }
     }
+
+    private boolean isRootCategoryPresent(final Collection<NodeRef> categoryNodeRefs)
+    {
+        return categoryNodeRefs.stream().anyMatch(this::isRootCategory);
+    }
+
+    private boolean isCategoryAspectMissing(final NodeRef nodeRef)
+    {
+        return !nodeService.hasAspect(nodeRef, ContentModel.ASPECT_GEN_CLASSIFIABLE);
+    }
+
+    /**
+     * Merge already present and new categories ignoring repeating ones.
+     *
+     * @param currentCategories Already present categories.
+     * @param newCategories Categories which should be added.
+     * @return Merged categories.
+     */
+    private Collection<NodeRef> mergeCategories(final Serializable currentCategories, final Collection<NodeRef> newCategories)
+    {
+        if (currentCategories == null)
+        {
+            return newCategories;
+        }
+
+        final Collection<NodeRef> allCategories = DefaultTypeConverter.INSTANCE.getCollection(NodeRef.class, currentCategories);
+        newCategories.stream()
+            .filter(category -> !allCategories.contains(category))
+            .forEach(allCategories::add);
+
+        return allCategories;
+    }
+
+    /**
+     * Add to or update node's property cm:categories containing linked category references.
+     *
+     * @param nodeRef Node reference.
+     * @param categoryNodeRefs Category node references.
+     */
+    private void linkNodeToCategories(final NodeRef nodeRef, final Collection<NodeRef> categoryNodeRefs)
```

Comment:
This looks fine to me, but can't we reuse some existing service code for this? (Maybe in the CategoryService?)

Reply:
Ok - that sounds like a good idea, but if it's a more general change then we should get this PR merged first.

Reply:
As Maciej already mentioned, unfortunately there doesn't exist any service code which could be reused. Similar solution is used in `LinkCategoryActionExecuter` and we could update it also there if we would introduce some new "CategoryService".